keps/sig-network/2433-topology-aware-hints/README.md

```diff
@@ -0,0 +1,578 @@
+  scope) this will not self-rectify.
+- Autoscaling will not behave well if only a single zone is receiving large
+  amounts of traffic. This could potentially be mitigated by separating
+  deployments and HPAs per zone.
+- Services with ExternalTrafficPolicy=local will need special treatment here.
+  This approach could result in a situation where an endpoint on a Node is
+  delivered to a separate underprovisioned zone. The simplest approach would be
+  to disable this functionality altogether.
+- When this feature is transitioning between enabled and disabled states, there
+  will be a brief point in time where only some EndpointSlices have hints. That
+  could temporarily result in traffic being routed to a small subset of
+  endpoints. To avoid this, we only filter out endpoints that have a hint set
+  to a different zone. If a hint is not set for an endpoint, it will be included
+  by all instances of kube-proxy.
+
+## Design Details
+
+### Assumptions
+
+- Incoming traffic is proportional to the number of allocatable CPU cores in a
+  zone. Although this is an imperfect metric, it is the best available way of
+  predicting how much traffic will be received in a zone. If we are unable to
+  derive the number of allocatable cores in a zone we will fall back to the
+  number of nodes in that zone.
+- Service capacity is proportional to the number of endpoints in a zone. This
+  assumes that each endpoint has equivalent capacity. Although this is not
+  always true, it usually is. We can explore ways to deal with variable capacity
+  endpoints in the future.
+
+### Configuration
+
+A new `topology.kubernetes.io/topology-aware-routing` annotation will be
+supported on Services. This will have 2 potential values:
+
+- `Auto`: When there are a sufficient number of endpoints for the Service, the
+  EndpointSlice controller will add topology hints for each endpoint that will
+  ensure a proportional amounts are available to each zone in a cluster.
+- `Disabled`: EndpointSlice hints will not be set for this feature.
+
+When this annotation is either unspecified or not set to `Auto`, the `Disabled`
+behavior will be default. A future KEP will explore changing the default
+behavior from `Disabled` to `Auto` with a new feature gate.
+
+#### Interoperability
+
+If both this annotation and `topologyKeys` are set on a Service, `topologyKeys`
+will be given precedence and this annotation will be ignored.
+
+#### Feature Gate
+
+This functionality will be guarded by the `TopologyAwareHints` feature
+gate. This gate should not be enabled if the `ServiceTopology` gate is enabled.
+
+### API
+
+A new `TopologyHints` struct would be added to the `EndpointSlice.Endpoint`
+struct:
+
+```go
+type Endpoint struct {
+  ...
+  // hints contains information associated with how an endpoint should be
+  // consumed.
+  // +optional
+  Hints EndpointHints `json:"hints,omitempty" protobuf:"bytes,7,opt,name=hints"`
```

Comment:
`hints` is very general - is it JUST for topology or can we imagine other hints?  E.g. might capacity end up here?

If we keep this name general, perhaps `zones` (or `zones`) should be `forZone` or `zoneMap`, or maybe it should be in a `usage` sub-struct:

```
hints:
  usage:
    zones;
    - zone: z1
    - zone: z2
``` 
?

Reply:
I think I like `forZones` here to avoid an extra level of nesting. Agree that this mechanism could be useful beyond topology.

Reply:
Updated to use `forZones`, also happy to add `usage` if that's preferred.